slave-base/contrib/bin/run-jnlp-client

```diff
@@ -65,57 +63,88 @@ if [[ $# -gt 1 ]]; then
+
+    if [[ -z "$CONTAINER_HEAP_PERCENT" ]]; then
         CONTAINER_HEAP_PERCENT=0.50
     fi
 
-    CONTAINER_MEMORY_IN_MB=$((${CONTAINER_MEMORY_IN_BYTES}/1024**2))
-    #if machine has 4GB or less, meaning max heap of 2GB given current default, force use of 32bit to save space unless user
-    #specifically want to force 64bit
-    HEAP_LIMIT_FOR_32BIT=$((2**32-1))
-    HEAP_LIMIT_FOR_32BIT_IN_MB=$((${HEAP_LIMIT_FOR_32BIT}/1024**2))
     CONTAINER_HEAP_MAX=$(echo "${CONTAINER_MEMORY_IN_MB} ${CONTAINER_HEAP_PERCENT}" | awk '{ printf "%d", $1 * $2 }')
-    if [[ -z $OPENSHIFT_JENKINS_JVM_ARCH && "${CONTAINER_HEAP_MAX}" -lt "${HEAP_LIMIT_FOR_32BIT_IN_MB}"  ]]; then
-      echo "max heap in MB is ${CONTAINER_HEAP_MAX} and 64 bit was not explicitly set so using 32 bit Java"
-      alternatives --set java $JVMPath32bit
-      export MALLOC_ARENA_MAX=1
+    if [[ $CONTAINER_HEAP_LIMIT_IN_MB && $CONTAINER_HEAP_MAX -gt $CONTAINER_HEAP_LIMIT_IN_MB ]]; then
+      CONTAINER_HEAP_MAX=$CONTAINER_HEAP_LIMIT_IN_MB
+    fi
+    if [[ -z "$JAVA_MAX_HEAP_PARAM" ]]; then
+      JAVA_MAX_HEAP_PARAM="-Xmx${CONTAINER_HEAP_MAX}m"
     fi
 
-    JAVA_MAX_HEAP_PARAM="-Xmx${CONTAINER_HEAP_MAX}m"
-    if [ -z $CONTAINER_INITIAL_PERCENT ]; then
-      CONTAINER_INITIAL_PERCENT=0.07
+    if [[ "$CONTAINER_INITIAL_PERCENT" ]]; then
+      CONTAINER_INITIAL_HEAP=$(echo "${CONTAINER_HEAP_MAX} ${CONTAINER_INITIAL_PERCENT}" | awk '{ printf "%d", $1 * $2 }')
+      if [[ -z "$JAVA_INITIAL_HEAP_PARAM" ]]; then
+        JAVA_INITIAL_HEAP_PARAM="-Xms${CONTAINER_INITIAL_HEAP}m"
+      fi
+    fi
+
+    # if the calculated max heap is < 2GiB and OPENSHIFT_JENKINS_JVM_ARCH is not
+    # set, use 32 bit JVM for space efficiency
+    if [[ -z $OPENSHIFT_JENKINS_JVM_ARCH ]]; then
+      if [[ "${CONTAINER_HEAP_MAX}" -gt 2048 ]]; then
+        echo "max heap in MB is ${CONTAINER_HEAP_MAX}, using 64 bit Java"
+        alternatives --set java $JVMPath64bit
+      else
+        echo "max heap in MB is ${CONTAINER_HEAP_MAX} and 64 bit was not explicitly set so using 32 bit Java"
+        alternatives --set java $JVMPath32bit
+        export MALLOC_ARENA_MAX=${MALLOC_ARENA_MAX:-1}
+      fi
     fi
-    CONTAINER_INITIAL_HEAP=$(echo "${CONTAINER_HEAP_MAX} ${CONTAINER_INITIAL_PERCENT}" | awk '{ printf "%d", $1 * $2 }')
-    JAVA_INITIAL_HEAP_PARAM="-Xms${CONTAINER_INITIAL_HEAP}m"
-  fi
 
-  if [ -z $JAVA_GC_OPTS ]; then
-      # We no longer set MaxMetaspaceSize because the JVM should expand metaspace until it reaches the container limit.
-      # See http://hg.openjdk.java.net/jdk8u/jdk8u/hotspot/file/4dd24f4ca140/src/share/vm/memory/metaspace.cpp#l1470
-      JAVA_GC_OPTS="-XX:+UseParallelGC -XX:MinHeapFreeRatio=5 -XX:MaxHeapFreeRatio=10 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90"
+  else
+    # no k8s/docker memory limit set
+    if [[ -z $OPENSHIFT_JENKINS_JVM_ARCH ]]; then
+      echo "no cgroup memory limit found, using 64 bit Java"
```

Comment:
OPENSHIFT_JENKINS_JVM_ARCH is not set and no cgroup memory limit was found, using 64 bit Java"